java/vector/src/main/java/org/apache/arrow/vector/ElementAddressableVector.java

```diff
@@ -0,0 +1,42 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.arrow.vector;
+
+import org.apache.arrow.memory.util.ArrowBufPointer;
+
+/**
+ * Vector for which each data element resides in a continuous memory region,
+ * so it can be pointed to by an {@link org.apache.arrow.memory.util.ArrowBufPointer}.
+ */
+public interface ElementAddressableVector extends ValueVector {
```

Comment:
why is a new interface needed?

Reply:
Currently, getDataPointer methods are supported by FixedWidthVector and VariableWidthVector interfaces. To implement the iterator properly, we need a common interface with the method. 

However, the common interface should not be ValueVector, because some sub-class should not support the getDataPointer operator, as their elements are not stored in a continuous memory region.

Reply:
Hmm, it seems like a better name for this might be PrimitiveVector?  All primitive types should be able to have a direct pointer to there elements?

Reply:
-0.9

I'm really struggling with this being added to the interface. I generally don't think this is a performant interface. It is one thing to be external to the core apis. Adding into the core APIs seems like it suggests that this is a good way to access data. For example, if you want to make a performant hash implementation, you wouldn't use this interface. What are we really solving with ArrowBufPointer? I feel like we're letting traditional java-isms leak into a codebase that was trying to avoid them for the most part.

Reply:
To add some detail, if you're reading 4 byte values, you should be working with ints in java, not a wrapper object that points to a byte buffer with an offset and length. A huge part of the performance you can get with Arrow is by using known structures: most are 1,2,4,8,16 byte widths. Writing things for each of those widths is way better than trying to layer a java object interface on top.